src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs

```diff
@@ -17,235 +18,247 @@ namespace System.Text.Json.Serialization.Tests
     public static partial class StreamTests
```

Comment:
Aside from applying review feedback, changes here are mostly cosmetic to make the code a bit cleaner & easier to read.

I also remove asserts on serialized dictionaries with `ReferenceHandling.Preserve` since reflection ordering is not guaranteed -- `ref`/`id` metadata values might not match up.

Reply:
> reflection ordering is not guaranteed -- ref/id metadata values might not match up.

Hmm, what does reflection ordering has to do with dictionary KeyValuePairs and metadata? 
Doesn't reflection ordering only affects the order of object's properties?

Reply:
I meant the order of elements returned by `.GetEnumerator()` for dictionaries, which is the order we serialize & check for metadata. The order is not guaranteed IIRC. I hit a few test failures because of this locally.

Reply:
> which is the order we serialize & check for metadata.

We don't check for metadata on serialization, we append metadata to the resulting JSON objects and collections; metadata properties are never part of the CLR object.

Reply:
Could it be that the issue that you are facing is caused by a race condition bug in the test?

Reply:
> > which is the order we serialize & check for metadata.
> 
> We don't check for metadata on serialization

check for **references***

We check if we have seen the instance before in order to assign it metadata, and that is done in the order returned by the enumerator.

The dictionary element that takes an "id" of "1" when we serialize may not be the same one that takes it during the next serialization.

> Could it be that the issue that you are facing is caused by a race condition bug in the test?

I don't believe so.

Reply:
As we discussed offline, a Dictionary enumeration returns elements in a non-deterministic manner on Serialization time, which will cause that its elements don't get the same id on each call. So condition on line 99 `if(!(DictionaryTypes<TElement>().Contains(type) && options.ReferenceHandling == ReferenceHandling.Preserve))` is correct.

Reply:
After this PR is merged, I'll open an issue discussing ways we can run the test for the `if(DictionaryTypes<TElement>().Contains(type) && options.ReferenceHandling == ReferenceHandling.Preserve)` case. One way is to override the GetEnumerator method and force a particular order.